### ISSUE TITLE
Switch to non-zlib based deflate compression?

### ISSUE THREAD

**Issue by thomaseizinger** [twittner/twist#72]
Currently, we unconditionally activate the `zlib` feature of the `flate2` crate. `flate2` also supports a pure Rust backend which is actually the default.

Can we switch to that and thus avoid relying on the shared library?

**Comment by ggwpez** [comment]
cc @niklasad1

**Comment by niklasad1** [comment]
IIRC,  `zlib` was preferred because it is faster than the pure rust backend. I get that it may be annoying for the C dependencies...

For now, I would prefer if we could expose a couple of feature flags to configure the backends for flate2 crate and keep zlib as default.

I have no good intuition how much faster zlib so better keep current behavior and if/when we get some numbers I'm fine switching to the rust backend as default.

**Comment by thomaseizinger** [comment]
Thanks for responding!

One of the issues with feature-flags is that the application which actually has to the deal with the shared library may not have access to your feature-flags because your library is 3 or 4 layers deep in the dependency tree.

Moreover, they are additive and thus can clash with other dependencies in the tree activating some of them already.

As far as I know, the best way to handle this is to use "bare" `cfg`s and not feature-flags.

In this case, `flate2` has already made the decision to use feature-flags meaning you cannot use `cfg`s.

However, the design they went for allows users to _force_ the use of `zlib` through a feature but that cannot be undone within a single dependency tree.

Therefore, my argument would be that forcing `zlib` onto users by activating it unconditionally in `soketto` is a mistake. If you want to use `zlib` in your app that uses soketto, you should depend on `flate2` yourself and enable the `zlib` feature flag. Everyone else just gets the pure Rust backend.

I am happy to send a PR to make the necessary changes.

**Comment by niklasad1** [comment]
> Therefore, my argument would be that forcing zlib onto users by activating it unconditionally in soketto is a mistake. If you want to use zlib in your app that uses soketto, you should depend on flate2 yourself and enable the zlib feature flag. Everyone else just gets the pure Rust backend.

> I am happy to send a PR to make the necessary changes.

In general, I think this approach is hard to understand for downstream users to understand and needs to be documented in some way but in this case it's probably best solution.

To elaborate soketto will actually rely on https://github.com/rust-lang/flate2-rs/blob/main/src/ffi/mod.rs#L37-#L46 to ensure that `--features zlib` overrides the rust-backend. It's a bit tricky whether changing that in the future is considering a breaking change or not.

I'm okay with this, @jsdw @lexnv what do you think?

**Comment by thomaseizinger** [comment]
> > Therefore, my argument would be that forcing zlib onto users by activating it unconditionally in soketto is a mistake. If you want to use zlib in your app that uses soketto, you should depend on flate2 yourself and enable the zlib feature flag. Everyone else just gets the pure Rust backend.
> 
> > I am happy to send a PR to make the necessary changes.
> 
> In general, I think this approach is hard to understand for downstream users to understand and needs to be documented in some way

Agreed, it is not easy to wrap one's head around the entire feature design of cargo. It is fairly novel from what I know so people cannot relate to it from other languages. I do think education on the "proper" solution is the best we can do here :)

> but in this case it's probably best solution.

Great, thank you!

> To elaborate soketto will actually rely on https://github.com/rust-lang/flate2-rs/blob/main/src/ffi/mod.rs#L37-#L46 to ensure that `--features zlib` overrides the rust-backend. It's a bit tricky whether changing that in the future is considering a breaking change or not.

You mean whether `flate2` will change their approach? It is a crate under the `rust-lang` org. I trust them to weigh off tradeoffs very carefully :)

**Comment by niklasad1** [comment]
> You mean whether flate2 will change their approach? It is a crate under the rust-lang org. I trust them to weigh off tradeoffs very carefully :)

Yeah my comment was more "general about features" but sure I agree :)

**Comment by jsdw** [comment]
So the idea here is to not unconditionally enable `zlib` in soketto and instead expect users who want it to depend on `flate2` themselves to enable the flag. Sounds fair to me!

**Comment by thomaseizinger** [comment]
So it turns out the reason we are depending on zlib isn't performance but because the Rust implementation doesn't support setting the window-bits: https://github.com/rust-lang/flate2-rs/issues/239#issuecomment-631674396
